pallets/compliance-manager/src/lib.rs

```diff
@@ -341,23 +333,28 @@ decl_module! {
             <<T as IdentityTrait>::ProtocolFee>::charge_fee(
                 ProtocolOp::ComplianceManagerAddComplianceRequirement
             )?;
-            let new_requirement = ComplianceRequirement {
+            let mut new_req = ComplianceRequirement {
                 sender_conditions,
                 receiver_conditions,
-                id: Self::get_latest_requirement_id(ticker) + 1u32
+                id: LastRequirementId::get(ticker) + 1u32
             };
+            new_req.dedup();
 
-            let mut asset_compliance = AssetCompliances::get(ticker);
-            let reqs = &mut asset_compliance.requirements;
-
+            let mut reqs = AssetCompliances::iter_prefix_values(ticker).collect::<Vec<_>>();
             if !reqs
                 .iter()
-                .any(|requirement| requirement.sender_conditions == new_requirement.sender_conditions && requirement.receiver_conditions == new_requirement.receiver_conditions)
+                .any(|req|
+                    req.sender_conditions == new_req.sender_conditions
+                    && req.receiver_conditions == new_req.receiver_conditions
+                )
             {
```

Comment:
This is unfortunate, as it prevents us from memoizing the complexity calculation for other requirements, so we have no choice but to fetch them. This renders adding a new requirement O(n) where `n` is the number of requirements for the ticker.

Reply:
tbh I'm not a fan of such data sanity checks onchain. Even if these checks fail, nothing bad is gonna happen. These checks should be done at the UI level IMO. That being said, I think we are ok with keeping this as it is for now, Due to our rule complexity requirement, this array will not grow much.